### PR TITLE
Fix the parent of a reference being undefined when the parent has never been accessed before

### PR DESCRIPTION
**Pull request by xaviergonz**
Fix for #993

### CODE REVIEW

**Comment by xaviergonz** [comment]
Please review, I'm not sure if this is the best approach

**Comment by xaviergonz** [comment]
actually, should I call. value from parent to child to ensure the creation order is done that way?

**Comment by xaviergonz** [comment]
or maybe I just need to make getParent and family use `.value` before `. storedValue` to force initialization, hmm 🤔

**Comment by k-g-a** [comment]
Hi! The fix seems efficient and elegant to me.
I only have doubt about parent.storedValue check in while. Could that be the case, when one of the intermediate parents is already initialized, while another parent upper the tree is not?
Maybe we should do:
```javascript
              if (matched) {
                    let parent = matched.parent
                    while (parent) {
                         if(!parent.storedValue) {
                          // accessing the value will ensure the node is initialized
                          // tslint:disable-next-line:no-unused-expression
                           parent.value
                         }
                         parent = parent.parent
                    }
                }
```

Another suggestion is to make ```_observableInstanceCreated```/```_createObservableInstance``` public and use them? So we make code self-documented and avoid need of comments like "accessing the value will ensure the node is initialized".

**Comment by xaviergonz** [comment]
@k-g-a I put the check in place to avoid traversing the tree up to the root everytime a reference is used, so I'd like to keep that, but of course only if that covers all cases.

so about the cases, you think there might be cases where a parent might be initialized but it's parent won't? (skipping the current bug of course)

also about afterCreation/afterAttach order, should the node initialization be done from root to child or from child to root? or it doesn't matter.

another thing I'd like to try is fixing instead getParent so it initializes the node if not done already (including the whole chain to the root) rather than at reference time but the same doubt about initialization remains

**Comment by xaviergonz** [comment]
ok, I just made the sure the order of the afterCreate/afterAttach events are the same when using a reference and the standard way

I still didn't remove the optimization though since I can't think of a way to access a node other than drilling through the store (l1.l2.l3.l4) or using a reference, (l1.selectedL4, which in turn initializes with the fix l2.l3.l4), but if you have any other ideas of when that could happen I will make sure to remove the optimization

File: packages/mobx-state-tree/__tests__/core/object.test.ts [hunk 1]
```diff
@@ -706,3 +707,115 @@ if (process.env.NODE_ENV === "development")
...
+
+        l1.selected.toggle()
+        l1.l2.l3.l4.toggle()
+        expect(events).toEqual(expectedEvents)
+    }
+
+    // test 3, reference get parent should be available from the beginning
+    {
+        const l1 = createL1()
+        expect(getParent(l1.selected)).toBeTruthy()
```
Comment (mweststrate) [inline packages/mobx-state-tree/__tests__/core/object.test.ts]: Note that this returns the parent of the selected's target, not (so L4's parent)

**Comment by mweststrate** [comment]
I am now wondering, can this issue also occur when using `getParent` in after attach?

File: packages/mobx-state-tree/__tests__/core/object.test.ts (hunk 1 above)
Comment (xaviergonz) [reply packages/mobx-state-tree/__tests__/core/object.test.ts]: yep, but if I change it to test against l3, then that reference will initialize it already, so the unit test wouldn't do much

Comment (xaviergonz) [reply packages/mobx-state-tree/__tests__/core/object.test.ts]: I'm not sure if I follow about the after attach and getParent thing...

**Comment by xaviergonz** [comment]
@mweststrate improved the unit test for getParent and added a unit test for after attach to check if the parent is there when it should be

**Comment by mweststrate** [comment]
@xaviergonz looking good! Approved

**Pull request merged**